pkg/controller/contrailcni/contrailcni_controller.go

```diff
@@ -0,0 +1,171 @@
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("contrailcni-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource ContrailCNI
+	err = c.Watch(&source.Kind{Type: &v1alpha1.ContrailCNI{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Nodes
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(nodeObject handler.MapObject) []reconcile.Request {
+			var cniObjects v1alpha1.ContrailCNIList
+			_ = mgr.GetClient().List(context.TODO(), &cniObjects)
+			var requests = []reconcile.Request{}
+			for _, cniObject := range cniObjects.Items {
+				requests = append(requests, reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      cniObject.Name,
+						Namespace: cniObject.Namespace,
+					},
+				})
+			}
+			return requests
+		}),
+	})
+	return err
+}
+
+// blank assignment to verify that ReconcileContrailCNI implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileContrailCNI{}
+
+// ReconcileContrailCNI reconciles a ContrailCNI object
+type ReconcileContrailCNI struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	Client      client.Client
+	Scheme      *runtime.Scheme
+	kubernetes  *k8s.Kubernetes
+	ClusterInfo v1alpha1.CNIClusterInfo
+}
+
+// Reconcile reads that state of the cluster for a ContrailCNI object and makes changes based on the state read
+// and what is in the ContrailCNI.Spec
+func (r *ReconcileContrailCNI) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ContrailCNI")
+	instanceType := "contrailcni"
+	instance := &v1alpha1.ContrailCNI{}
+	ctx := context.TODO()
+
+	if err := r.Client.Get(ctx, request.NamespacedName, instance); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if !instance.GetDeletionTimestamp().IsZero() {
+		return reconcile.Result{}, nil
+	}
+
+	if instance.Spec.ServiceConfiguration.ControlInstance != "" {
+		controlInstance := v1alpha1.Control{}
+		configInstance := v1alpha1.Config{}
+		configActive := configInstance.IsActive(instance.Labels["contrail_cluster"],
+			request.Namespace, r.Client)
+
+		controlActive := controlInstance.IsActive(instance.Spec.ServiceConfiguration.ControlInstance,
+			request.Namespace, r.Client)
+
+		if !configActive || !controlActive {
+			return reconcile.Result{}, nil
+		}
```

Comment:
Because there's difference in implementations of this methods for config and control. Probably because control is defined as list and config as signgle resource. One method gathers config by label selector while control is gathered by name.